[Merged by Bors] - Kiln mev boost

realbigsean opened the pull request:
## Issue Addressed

MEV boost compatibility

## Proposed Changes

See #2987

## Additional Info

This is blocked on the stabilization of a couple specs, [here](https://github.com/ethereum/beacon-APIs/pull/194) and [here](https://github.com/flashbots/mev-boost/pull/20).

Additional TODO's and outstanding questions

- [ ] MEV boost JWT Auth
- [ ] Will `builder_proposeBlindedBlock` return the revealed payload for the BN to propogate
- [ ] Should we remove `private-tx-proposals` flag and communicate BN <> VC with blinded blocks by default once these endpoints enter the beacon-API's repo? This simplifies merge transition logic.

realbigsean committed:
- VC changes to support private tx block signing
- `PrivateBeaconBlock`
- `PrivateBeaconBlockBody`
- `SignedPrivateBeaconBlock`

realbigsean committed:
- BN changes to support private tx block signing
- GET lighthouse/validator/blocks_private endpoint

realbigsean committed:
- POST lighthouse/beacon/blocks_private endpoint -- stubbed out

realbigsean committed:
rebase onto kintsugi and add payload builders to execution layer

realbigsean committed:
add blinded beacon block proposals

realbigsean committed:
rename private block to blinded block

realbigsean committed:
payload builder flag

realbigsean committed:
fix flag and query the correct nodes for the terminal block

realbigsean committed:
avoid TTD query

realbigsean committed:
Revert "avoid TTD query"

This reverts commit 5982466c63d3ef6a0f688b3975a1e1db33a37798.

realbigsean committed:
set ttd to zero and only try to submit private blocks after transition

realbigsean committed:
Fixes and updates

realbigsean committed:
Fix propose blinded block expected response

realbigsean committed:
attempt at making transactions generic in execution payloads

realbigsean committed:
fix call to EL from beacon http server

realbigsean committed:
gossip block received from mev boost

realbigsean committed:
Make transactions generic in an execution payload

realbigsean committed:
Cleanup, fixes

realbigsean committed:
convert blinded and non-blinded blocks

realbigsean committed:
Fixes

realbigsean committed:
remove some unused imports

realbigsean committed:
delete unused imports

realbigsean committed:
cleanup

realbigsean committed:
delete some comments

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into generic-transactions

 Conflicts:
	Cargo.lock
	consensus/state_processing/src/per_block_processing.rs
	consensus/state_processing/src/per_block_processing/block_signature_verifier.rs
	consensus/types/src/beacon_block.rs
	consensus/types/src/beacon_block_body.rs
	consensus/types/src/execution_payload.rs
	testing/ef_tests/src/cases/operations.rs

realbigsean committed:
update superstruct version

realbigsean committed:
remove serde_transactions

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into generic-transactions

realbigsean committed:
bellatrix rename fix

realbigsean committed:
add back advanced-pre-releases.md with the correct link

realbigsean committed:
add back advanced-pre-releases.md with the correct link

realbigsean committed:
improve fallback functionality, allow multiple payload builders

realbigsean committed:
fix log

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into generic-transactions

 Conflicts:
	beacon_node/http_api/src/lib.rs
	consensus/types/Cargo.toml

realbigsean committed:
lints and tests

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into kiln-mev-boost

 Conflicts:
	beacon_node/beacon_chain/src/beacon_chain.rs
	beacon_node/beacon_chain/src/execution_payload.rs
	beacon_node/execution_layer/src/engine_api/http.rs
	beacon_node/execution_layer/src/lib.rs
	beacon_node/execution_layer/src/test_utils/handle_rpc.rs
	beacon_node/execution_layer/src/test_utils/mock_execution_layer.rs
	beacon_node/http_api/Cargo.toml
	beacon_node/http_api/src/lib.rs
	beacon_node/lighthouse_network/Cargo.toml
	beacon_node/src/config.rs
	validator_client/src/validator_store.rs

realbigsean committed:
merge with unstable

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into kiln-mev-boost

 Conflicts:
	beacon_node/beacon_chain/src/test_utils.rs
	beacon_node/client/src/builder.rs
	beacon_node/client/src/config.rs
	beacon_node/execution_layer/src/engine_api.rs
	beacon_node/execution_layer/src/engine_api/http.rs
	beacon_node/execution_layer/src/engine_api/json_structures.rs
	beacon_node/execution_layer/src/lib.rs
	beacon_node/execution_layer/src/test_utils/execution_block_generator.rs
	beacon_node/execution_layer/src/test_utils/mock_execution_layer.rs
	beacon_node/src/config.rs
	consensus/types/src/execution_payload.rs
	consensus/types/src/lib.rs
	testing/execution_engine_integration/src/test_rig.rs

realbigsean commented on beacon_node/execution_layer/src/engine_api/json_structures.rs:
```diff
@@ -400,75 +524,3 @@ pub mod serde_logs_bloom {
             .map_err(|e| serde::de::Error::custom(format!("invalid logs bloom: {:?}", e)))
     }
 }
-
```
This is no longer necessary since we removed the union from the transaction type in the consensus spec

michaelsproul replied:
Nice

realbigsean committed:
cargo fmt and small cleanup

realbigsean committed:
create transactions.rs

realbigsean committed:
cargo fmt cargo fix

realbigsean committed:
update endpoints to match beacon-apis spec

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into kiln-mev-boost

 Conflicts:
	beacon_node/execution_layer/src/lib.rs
	testing/execution_engine_integration/src/test_rig.rs

realbigsean committed:
fix versioning on validator blinded block endpoint

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -535,64 +568,121 @@ impl ExecutionLayer {
-                    .get_payload_id(parent_hash, timestamp, prev_randao, suggested_fee_recipient)
+        match Txns::block_type() {
+            BlockType::Blinded => {
+                debug!(
+                    self.log(),
+                    "Issuing engine_getPayloadHeader";
+                    "suggested_fee_recipient" => ?suggested_fee_recipient,
+                    "prev_randao" => ?prev_randao,
+                    "timestamp" => timestamp,
+                    "parent_hash" => ?parent_hash,
+                );
```
I think this log message should be lifted out so that it occurs in both `Blinded`/`Full` cases

realbigsean replied:
Whoops there's a couple mistakes here.. I accidentally deleted the full block log here, probably during a merge with unstable. And the two logs are slightly different because the endpoints they hit have different names. This one should be `builder_getPayloadHeader` and the other one should be `engine_getPayload`

michaelsproul commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1023,6 +1024,119 @@ pub fn serve<T: BeaconChainTypes>(
+                                    e
+                                ))
+                            })?;
+                        let new_block = SignedBeaconBlock::Merge(SignedBeaconBlockMerge {
```
Just making a note here that we only support merge blocks on this endpoint currently, but the spec as currently written seems to imply we should support pre-Bellatrix blocks too

michaelsproul replied:
This is tracked by #3091

michaelsproul commented on beacon_node/execution_layer/src/engine_api/http.rs:
```diff
@@ -216,6 +224,115 @@ impl EngineApi for HttpJsonRpc {
+        &self,
+        block: SignedBeaconBlock<T, BlindedTransactions>,
+    ) -> Result<ExecutionPayload<T>, Error> {
+        let blinded_block = SignedBlindedBeaconBlock::from(block);
```
Can we not use the `SignedBeaconBlock<T, BlindedTransactions>` type directly here, and do away with the separate `BlindedBeaconBlock` types? Is there something about (de)serialization that breaks?

realbigsean replied:
The issue I ran into was the name of the field `transactions` (signed block's execution payload) vs `transactions_root` (signed blinded block's execution payload header). We could maybe figure out a way to do a conditional rename but I don't think there's a way to do it within `serde_derive`.  You can [alias](https://serde.rs/field-attrs.html#alias) on deserialization, but there's nothing like a `rename_if` for serialization.

This is actually making me realize we'd be slightly off-spec as-is because our BN<>VC communication would have the field name `transactions` if we don't do this same transformation in that API.

michaelsproul replied:
I've opened a PR to this branch that uses `serde(flatten)` to unify the types, let me know if you think it's flexible enough: https://github.com/realbigsean/lighthouse/pull/20

michaelsproul commented on validator_client/src/validator_store.rs:
```diff
@@ -5,6 +5,7 @@ use crate::{
     signing_method::{Error as SigningError, SignableMessage, SigningContext, SigningMethod},
 };
 use account_utils::{validator_definitions::ValidatorDefinition, ZeroizeString};
+use eth2::types::Transactions;
```
we may as well import this through `types::Transactions`

michaelsproul commented on validator_client/src/block_service.rs:
```diff
@@ -202,16 +235,48 @@ impl<T: SlotClock + 'static, E: EthSpec> BlockService<T, E> {
+            .bellatrix_fork_epoch
+            .unwrap_or_else(Epoch::max_value)
+            .start_slot(E::slots_per_epoch());
+        let now = self.slot_clock.now().unwrap();
```
I think we can use `slot` from further back in the function and avoid the `unwrap`

michaelsproul commented on validator_client/src/block_service.rs:
```diff
@@ -202,16 +235,48 @@ impl<T: SlotClock + 'static, E: EthSpec> BlockService<T, E> {
             self.inner.context.executor.spawn(
                 async move {
-                    if let Err(e) = service.publish_block(slot, validator_pubkey).await {
+                    //TODO(sean): how should pre-transition and the transition be handled?
+                    let publish_result = if private_tx_proposals && now > merge_slot {
```
this seems like a good way to handle the transition, although I would have thought the check should be `now >= merge_slot`?

what problems do you foresee?

realbigsean replied:
I was generally worried about outsourcing payload construction at transition, which paul addressed [here](https://github.com/sigp/lighthouse/pull/2987#issuecomment-1028487109) (I haven't implented his suggestion) 

But if this endpoint will end up serving pre-Bellatrix blocks, it might be better to remove any slot logic here and just add logic for "only outsource payload construction if we have finalized" to the beacon node

michaelsproul replied:
Ah yeah, that sounds sensible. I'm happy to come back to that after merging an MVP version of this PR.

michaelsproul reviewed (changes requested):
This is looking really neat!

I'd like to get this merged ASAP (post v2.1.4) so that we can start re-structuring the database around blinded blocks (i.e. store blinded blocks in the DB, with separate execution payloads that get erased on finalization).

I think so long as we have the types nailed down and we maintain backwards compatibility with the existing APIs when `--private-tx-proposals` is turned off then I think we can make further tweaks to the blinded block proposal flow after merging.

> Should we remove private-tx-proposals flag and communicate BN <> VC with blinded blocks by default once these endpoints enter the beacon-API's repo? This simplifies merge transition logic.

I think we should keep the opt-in flag so that the VC retains the ability to sign full blocks, and so we can maintain backwards compatibility with BNs that lack the `blinded_blocks` APIs leading up to the merge.

----

As far as I can tell the main changes that might occur can be handled with small tweaks to this PR after merging:

* Add support for pre-Bellatrix blocks to the `/blinded_blocks` HTTP endpoints
* Re-structure to support mev-boost as a "side car" rather than middleware, as per https://github.com/flashbots/mev-boost/pull/20/files#r788082526. I'm not sure what the appetite is for this change and whether it'll go through

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into kiln-mev-boost

realbigsean committed:
fix BN <> VC API field naming for blinded blocks

realbigsean committed:
PR feedback

realbigsean committed:
fix beacon chain tests

realbigsean committed:
cargo fmt

realbigsean committed:
remove unnecessary clone

realbigsean committed:
cargo fmt

realbigsean committed:
clean up fixme's

realbigsean commented:
Raised an issue [here](https://github.com/sigp/lighthouse/issues/3091) to track additional changes outside of this PR to try to get the new types merged more quickly

michaelsproul committed:
Remove separate Blinded types

realbigsean committed:
Merge pull request #20 from michaelsproul/kmb-flatten-payloads

Remove separate Blinded types

realbigsean committed:
remove any sync check queries to payload builders

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into kiln-mev-boost

realbigsean committed:
fix beacon chain tests

realbigsean committed:
serde skip phantom data

realbigsean commented:
@michaelsproul so I've tested this now and gotten it working using this repo: https://github.com/realbigsean/lighthouse-kiln-mev-boost

An outstanding limitation is that there is no relay running that can handle kiln blocks so I'm working on a local mock relay in a different branch. 

I had to make some updates to get this working here. The main thing is that mev-boost no longer handles any of our `eth_syncing` calls, so we can no longer incorporate it into our usual fallback routine, where we repeatedly make those calls and update a local picture of it's sync state.  I am going to make some larger changes in a separate PR, to separate the builder API from the engine API and untangle the fallback routine of the two. In this PR is a minimal, workable, not-ideal solution where I've just removed any `eth_syncing` calls but in doing so, removed builder to builder fallback logic. 

Additionally there's a couple of ssz static tests failing.. I'm thinking it might be because we need something like `#[ssz(flatten)]` for the new payload wrapper?

realbigsean commented:
Oh you know it's actually failing when it tries to read from the yaml file. I think `#[serde(flatten)]` and `#[serde(transparent)]` are both removing a layer of the struct hierarchy so we end up with `BeaconBlockBody` including `ExecutionPayload` fields directly

realbigsean committed:
remove `#[serde(transparent)]` from `Payload` structs

realbigsean commented:
Ok I think this should fix it https://github.com/sigp/lighthouse/pull/3062/commits/c36bafca4333edfa4c16a8953127264d22892999

🤞

michaelsproul commented:
Oh yeah, good catch. Looks like that fixed it!

realbigsean committed:
make engine_api::http crate public

realbigsean committed:
make `EngineApi` trait public

realbigsean committed:
just make all of engine_api public

realbigsean committed:
just make all of engine_api public

realbigsean commented:
I made some more parts of the `execution_layer` crate public for use in the mock relay I'm working on [here](https://github.com/realbigsean/mock-relay)

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into kiln-mev-boost

realbigsean committed:
- Only query builders with payload attributes populated
- make it possible for  jsonrpc id to be a string or an int

realbigsean committed:
builder and engine api separation

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into kiln-mev-boost

realbigsean committed:
Merge branch 'split-builder-engine-apis' of https://github.com/realbigsean/lighthouse into kiln-mev-boost

realbigsean committed:
fixes after merging engine builder separation

realbigsean committed:
generate lockfile

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into kiln-mev-boost

 Conflicts:
	beacon_node/beacon_chain/src/beacon_chain.rs
	beacon_node/http_api/src/lib.rs
	common/eth2/src/lib.rs

realbigsean committed:
merge updates

realbigsean committed:
fix tests

michaelsproul commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1022,6 +1023,116 @@ pub fn serve<T: BeaconChainTypes>(
+                                        .message()
+                                        .body()
+                                        .sync_aggregate()
+                                        .unwrap()
```
This `unwrap` isn't ideal, but I note that you've fixed it in #3134 and it isn't reachable from non-blinded block proposal routines, so I think we're good :+1:

realbigsean replied:
Whoops, oversight on my part!

michaelsproul reviewed (approved):
I'm happy to merge this!

michaelsproul commented:
Merging this so PRs built on top can target unstable

bors r+
